File: Python/ExampleTests/production/SimpleJson.py
"""The compact JSON-like text used by the Json specification: names are not
quoted, values are, and whitespace between tokens is insignificant.

    {anInt:"1",aString:"B"}
    [{anInt:"1",aString:"B"},{anInt:"2",aString:"C"}]

Field order is preserved, so a canonical form can be compared directly.
"""


class _Cursor:
    def __init__(self, text: str):
        self.text = text if text is not None else ''
        self.i = 0

    def at_end(self) -> bool:
        return self.i >= len(self.text)

    def peek(self) -> str:
        return '' if self.at_end() else self.text[self.i]

    def next(self) -> str:
        if self.at_end():
            return ''
        ch = self.text[self.i]
        self.i += 1
        return ch

    def skip_whitespace(self):
        while not self.at_end() and self.text[self.i].isspace():
            self.i += 1

    def expect(self, expected: str):
        self.skip_whitespace()
        if self.next() != expected:
            raise ValueError(f"Expected '{expected}' at {self.i}")


class SimpleJson:

    # -----------------------------------------------------------------
    # Writing
    # -----------------------------------------------------------------

    @staticmethod
    def to_object(fields: dict) -> str:
        parts = []
        for name, value in fields.items():
            parts.append(f'{name}:{SimpleJson._quoted(value)}')
        return '{' + ','.join(parts) + '}'

    @staticmethod
    def to_array(rows) -> str:
        return '[' + ','.join(SimpleJson.to_object(row) for row in rows) + ']'

    @staticmethod
    def _quoted(value) -> str:
        out = ['"']
        if value is not None:
            for c in str(value):
                if c in ('"', '\\'):
                    out.append('\\')
                out.append(c)
        out.append('"')
        return ''.join(out)

    # -----------------------------------------------------------------
    # Reading
    # -----------------------------------------------------------------

    @staticmethod
    def parse_object(text: str) -> dict:
        c = _Cursor(text)
        fields = SimpleJson._read_object(c)
        c.skip_whitespace()
        if not c.at_end():
            raise ValueError(f'Unexpected text after object at {c.i}')
        return fields

    @staticmethod
    def parse_array(text: str) -> list:
        c = _Cursor(text)
        c.skip_whitespace()
        c.expect('[')
        rows = []
        c.skip_whitespace()
        if c.peek() == ']':
            c.next()
        else:
            while True:
                rows.append(SimpleJson._read_object(c))
                c.skip_whitespace()
                d = c.next()
                if d == ',':
                    continue
                if d == ']':
                    break
                raise ValueError(f"Expected ',' or ']' at {c.i}")
        c.skip_whitespace()
        if not c.at_end():
            raise ValueError(f'Unexpected text after array at {c.i}')
        return rows

    @staticmethod
    def without_whitespace(text: str) -> str:
        """Remove whitespace that sits between tokens, leaving a plain string
        that can be compared to another one directly. Whitespace inside a
        quoted value is part of the value and is kept."""
        if text is None:
            return ''
        out = []
        in_quotes = False
        i = 0
        while i < len(text):
            c = text[i]
            if in_quotes:
                out.append(c)
                if c == '\\' and i + 1 < len(text):
                    out.append(text[i + 1])
                    i += 1
                elif c == '"':
                    in_quotes = False
            elif c == '"':
                in_quotes = True
                out.append(c)
            elif not c.isspace():
                out.append(c)
            i += 1
        return ''.join(out)

    # -----------------------------------------------------------------

    @staticmethod
    def _read_object(c: _Cursor) -> dict:
        c.skip_whitespace()
        c.expect('{')
        fields = {}
        c.skip_whitespace()
        if c.peek() == '}':
            c.next()
            return fields
        while True:
            c.skip_whitespace()
            name = SimpleJson._read_name(c)
            c.skip_whitespace()
            c.expect(':')
            c.skip_whitespace()
            fields[name] = SimpleJson._read_value(c)
            c.skip_whitespace()
            d = c.next()
            if d == ',':
                continue
            if d == '}':
                return fields
            raise ValueError(f"Expected ',' or '}}' at {c.i}")

    @staticmethod
    def _read_name(c: _Cursor) -> str:
        """A name is bare text up to the colon, or a quoted string."""
        if c.peek() == '"':
            return SimpleJson._read_quoted(c)
        out = []
        while not c.at_end() and c.peek() != ':':
            out.append(c.next())
        return ''.join(out).strip()

    @staticmethod
    def _read_value(c: _Cursor) -> str:
        """A value is a quoted string, or bare text up to the next ',' or '}'."""
        if c.peek() == '"':
            return SimpleJson._read_quoted(c)
        out = []
        while not c.at_end() and c.peek() not in (',', '}'):
            out.append(c.next())
        return ''.join(out).strip()

    @staticmethod
    def _read_quoted(c: _Cursor) -> str:
        c.expect('"')
        out = []
        while True:
            if c.at_end():
                raise ValueError(f'Unterminated string at {c.i}')
            ch = c.next()
            if ch == '"':
                return ''.join(out)
            if ch == '\\' and not c.at_end():
                ch = c.next()
            out.append(ch)

File: Swift/ExampleTests/production/SimpleJson.swift
import Foundation

/// The compact JSON-like text used by the Json specification: names are not
/// quoted, values are, and whitespace between tokens is insignificant.
///
///     {anInt:"1",aString:"B"}
///     [{anInt:"1",aString:"B"},{anInt:"2",aString:"C"}]
///
/// Field order is preserved, so a canonical form can be compared directly.
public enum SimpleJson {

    /// One name/value pair. An array of these keeps the declared order, which a
    /// dictionary would not.
    public typealias Field = (name: String, value: String)

    public struct ParseError: Error, CustomStringConvertible {
        public let message: String
        public init(_ message: String) { self.message = message }
        public var description: String { return message }
    }

    // -------------------------------------------------------------------------
    // Writing
    // -------------------------------------------------------------------------

    public static func toObject(_ fields: [Field]) -> String {
        let parts = fields.map { "\($0.name):\(quoted($0.value))" }
        return "{" + parts.joined(separator: ",") + "}"
    }

    public static func toArray(_ rows: [[Field]]) -> String {
        return "[" + rows.map { toObject($0) }.joined(separator: ",") + "]"
    }

    private static func quoted(_ value: String) -> String {
        var out = "\""
        for c in value {
            if c == "\"" || c == "\\" { out.append("\\") }
            out.append(c)
        }
        out.append("\"")
        return out
    }

    // -------------------------------------------------------------------------
    // Reading
    // -------------------------------------------------------------------------

    public static func parseObject(_ text: String) throws -> [Field] {
        var c = Cursor(text)
        let fields = try readObject(&c)
        c.skipWhitespace()
        guard c.atEnd else { throw ParseError("Unexpected text after object at \(c.pos)") }
        return fields
    }

    public static func parseArray(_ text: String) throws -> [[Field]] {
        var c = Cursor(text)
        c.skipWhitespace()
        try c.expect("[")
        var rows: [[Field]] = []
        c.skipWhitespace()
        if c.peek == "]" {
            _ = c.next()
        } else {
            while true {
                rows.append(try readObject(&c))
                c.skipWhitespace()
                let d = c.next()
                if d == "," { continue }
                if d == "]" { break }
                throw ParseError("Expected ',' or ']' at \(c.pos)")
            }
        }
        c.skipWhitespace()
        guard c.atEnd else { throw ParseError("Unexpected text after array at \(c.pos)") }
        return rows
    }

    /// Removes whitespace that sits between tokens, leaving a plain string that
    /// can be compared to another one directly. Whitespace inside a quoted value
    /// is part of the value and is kept.
    public static func withoutWhitespace(_ text: String) -> String {
        var out = ""
        var inQuotes = false
        var escaped = false
        for c in text {
            if inQuotes {
                out.append(c)
                if escaped {
                    escaped = false
                } else if c == "\\" {
                    escaped = true
                } else if c == "\"" {
                    inQuotes = false
                }
            } else if c == "\"" {
                inQuotes = true
                out.append(c)
            } else if !c.isWhitespace {
                out.append(c)
            }
        }
        return out
    }

    // -------------------------------------------------------------------------

    private static func readObject(_ c: inout Cursor) throws -> [Field] {
        c.skipWhitespace()
        try c.expect("{")
        var fields: [Field] = []
        c.skipWhitespace()
        if c.peek == "}" {
            _ = c.next()
            return fields
        }
        while true {
            c.skipWhitespace()
            let name = try readName(&c)
            c.skipWhitespace()
            try c.expect(":")
            c.skipWhitespace()
            fields.append((name: name, value: try readValue(&c)))
            c.skipWhitespace()
            let d = c.next()
            if d == "," { continue }
            if d == "}" { return fields }
            throw ParseError("Expected ',' or '}' at \(c.pos)")
        }
    }

    /// A name is bare text up to the colon, or a quoted string.
    private static func readName(_ c: inout Cursor) throws -> String {
        if c.peek == "\"" { return try readQuoted(&c) }
        var out = ""
        while !c.atEnd && c.peek != ":" { out.append(c.next()) }
        return out.trimmingCharacters(in: .whitespaces)
    }

    /// A value is a quoted string, or bare text up to the next ',' or '}'.
    private static func readValue(_ c: inout Cursor) throws -> String {
        if c.peek == "\"" { return try readQuoted(&c) }
        var out = ""
        while !c.atEnd && c.peek != "," && c.peek != "}" { out.append(c.next()) }
        return out.trimmingCharacters(in: .whitespaces)
    }

    private static func readQuoted(_ c: inout Cursor) throws -> String {
        try c.expect("\"")
        var out = ""
        while true {
            guard !c.atEnd else { throw ParseError("Unterminated string at \(c.pos)") }
            var ch = c.next()
            if ch == "\"" { return out }
            if ch == "\\" && !c.atEnd { ch = c.next() }
            out.append(ch)
        }
    }

    private struct Cursor {
        private let chars: [Character]
        private var i = 0

        init(_ text: String) { chars = Array(text) }

        var pos: Int { return i }
        var atEnd: Bool { return i >= chars.count }
        var peek: Character { return atEnd ? "\0" : chars[i] }

        mutating func next() -> Character {
            if atEnd { return "\0" }
            let c = chars[i]
            i += 1
            return c
        }

        mutating func skipWhitespace() {
            while !atEnd && chars[i].isWhitespace { i += 1 }
        }

        mutating func expect(_ expected: Character) throws {
            skipWhitespace()
            if next() != expected {
                throw ParseError("Expected '\(expected)' at \(i)")
            }
        }
    }
}
